src/main/scala/com/microsoft/hyperspace/index/dataskipping/DataSkippingIndex.scala

```diff
@@ -0,0 +1,180 @@
+      ctx: IndexerContext,
+      appendedSourceData: Option[DataFrame],
+      deletedSourceDataFiles: Seq[FileInfo],
+      indexContent: Content): (Index, Index.UpdateMode) = {
+    if (appendedSourceData.nonEmpty) {
+      writeImpl(ctx, index(ctx, appendedSourceData.get), SaveMode.Overwrite)
+    }
+    if (deletedSourceDataFiles.nonEmpty) {
+      val spark = ctx.spark
+      import spark.implicits._
+      val deletedFileIds = deletedSourceDataFiles.map(_.id).toDF(IndexConstants.DATA_FILE_NAME_ID)
+      val updatedIndexData = spark.read
+        .parquet(indexContent.files.map(_.toString): _*)
+        .join(deletedFileIds, Seq(IndexConstants.DATA_FILE_NAME_ID), "left_anti")
+      val writeMode = if (appendedSourceData.nonEmpty) {
+        SaveMode.Append
+      } else {
+        SaveMode.Overwrite
+      }
+      writeImpl(ctx, updatedIndexData, writeMode)
+    }
+    val updateMode = if (deletedSourceDataFiles.isEmpty) {
+      Index.UpdateMode.Merge
+    } else {
+      Index.UpdateMode.Overwrite
+    }
+    (this, updateMode)
+  }
+
+  override def refreshFull(
+      ctx: IndexerContext,
+      sourceData: DataFrame): (DataSkippingIndex, DataFrame) = {
+    val updatedIndex = copy(sketches = ExpressionUtils.resolve(ctx.spark, sketches, sourceData))
+    (updatedIndex, updatedIndex.index(ctx, sourceData))
+  }
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case DataSkippingIndex(thatSketches, _) => sketches.toSet == thatSketches.toSet
+      case _ => false
+    }
+
+  override def hashCode: Int = sketches.map(_.hashCode).sum
+
+  /**
+   * Creates index data for the given source data.
+   */
+  def index(ctx: IndexerContext, sourceData: DataFrame): DataFrame = {
+    val fileNameCol = "input_file_name"
+    val indexDataWithFileName = sourceData
+      .groupBy(input_file_name().as(fileNameCol))
+      .agg(aggregateFunctions.head, aggregateFunctions.tail: _*)
+
+    // Construct a dataframe to convert file names to file ids.
+    val spark = ctx.spark
+    val relation = RelationUtils.getRelation(spark, sourceData.queryExecution.optimizedPlan)
+    import spark.implicits._
+    val fileIdDf = ctx.fileIdTracker
+      .getIdToFileMapping(relation.pathNormalizer)
+      .toDF(IndexConstants.DATA_FILE_NAME_ID, fileNameCol)
+
+    indexDataWithFileName
+      .join(
+        fileIdDf.hint("broadcast"),
+        IndexUtils.decodeInputFileName(indexDataWithFileName(fileNameCol)) ===
+          fileIdDf(fileNameCol))
+      .select(
+        IndexConstants.DATA_FILE_NAME_ID,
+        indexDataWithFileName.columns.filterNot(_ == fileNameCol).map(c => s"`$c`"): _*)
+  }
+
+  private def writeImpl(ctx: IndexerContext, indexData: DataFrame, writeMode: SaveMode): Unit = {
+    indexData.cache()
+    indexData.count() // force cache
+    // Note: the actual index data size can be smaller due to compression.
+    val indexDataSize = indexData.queryExecution.optimizedPlan.stats.sizeInBytes
+    indexData.unpersist()
```

Comment:
Seems don't need to cache here; optimizedPlan.stats.sizeInBytes doesn't read from the cached data, but just estimate the plan.

Reply:
If you don't cache, the reported size is close to the size of the source data because it doesn't take the aggregations into account. The difference is huge.

Even if you cache, I found that optimizePlan.stats.sizeInBytes is somewhat too rough for the purpose. Please refer to the new commit.

Reply:
I understand your concern about `cache()`. I'll check the performance with some real data.

Reply:
@clee704  Can we use the max record limit or `spark.hyperspace.index.dataskipping.numPartitions` config? To me, these looks unnecessary complexity.
In most cases, DS index data file will be stored in 1 file. Even it's not evenly distributed, it's okay. We don't need to measure the exact size; repartition distributes rows by round robin and compression rate will be different. 

The reason I made the comment - https://github.com/microsoft/hyperspace/pull/461#discussion_r656790431 - is.. I just wanted to make it configurable.

Reply:
What is so complex about the code here? It's just a few lines of code. It's for the best user experience. If we fix the number of files with such a config, users should set the config value by themselves every time they create an index. Most of the time, they won't know the final index size and should create the index twice. If they set the value to 1 and don't change it, there is a possibility of OOM or slow write because write is done by a single task. Therefore, I cannot accept using such a config. It could be used as a supplementary config in addition to the current implementation, though.

I don't know where you are getting such strong concern for caching, but here are actual numbers:

Data: catalog_sales (from TPC-DS 1TB)
Stored size (parquet files): 138.4 GiB
Row count: 1,439,980,416
Number of files: 1,837

cache | sketch | # index data files | index creation time | index data size
--- | --- | --: | --: | --:
with cache | MinMaxSketch | 1 | 1m 4s | 30 KB
without cache | MinMaxSketch | 1 | 1m 2s | 30 KB
with cache | BloomFilterSketch | 4 | 1m 30s | 1 GB
without cache | BloomFilterSketch | 1 | 1m 40s | 1 GB
without cache | BloomFilterSketch | 4 | 1m 26s | 1 GB

"with cache" indicates the current implementation is used. "without cache" is repartition using the value from a config without caching.

As you can see, the overhead is minimal and less than 5% of the total creation time. The reality would be worse for the "without cache" case - you have to try multiple times to create an index because you don't know the index size, and always repartitioning to a single partition doesn't work for all workloads (OOM/too slow write).

If you still don't think caching isn't right, could you please state more clearly what's so bad about it?

Reply:
File scan is not done by a single task - spark creates multiple partitions to read a file based on partition size configs. (Partition num = task num)

Could you share the cluster size and result using a small cluster like 2 small nodes?

cc @imback82 @andrei-ionescu  @rapoth @paryoja Could you add some comment on this? 
I think cache() for the exact RDD byte size to determine the number of DS index files is unnecessary. Since in most cases it would be 1. 

So I suggested to use maxRecordsPerFile config or index.dataskipping.numPartitions
config which does not need an additional spark job and cache. 

Just my opinion and I'm not saying it's a must; I'd like to hear from other folks.

Reply:
The data was for 10 executors / 40 cores. Most of the index creation time comes from reading and aggregating the source data. Using a small cluster won't change the overall results regarding caching. But I'll give the results with 2 executors.

Index data sizes can range from a few KB to tens of GB or more, depending on the number of the source data files, types of sketches, number of distinct values for each column, etc. Repartition by 1 wouldn't work for large index data.

Reply:
There is another option: write the index data without any repartitioning, and check if the written index data files are too small (smaller than the configured target index data file size). If it's the case, read the files again to merge them. How about this?

Reply:
Thanks for the update! Without repartition, each file would usually be small. 
Seems there's no major concern from others, I'll approve the change.